keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/ArtifactInterfaces.kt

```diff
@@ -0,0 +1,82 @@
+package com.netflix.spinnaker.keel.api
+
+import com.netflix.spinnaker.keel.api.artifacts.ArtifactType
+
+/**
+ * Implemented by [ResourceSpec] or concrete resource types that (may) contain artifact information, typically compute
+ * resources.
+ *
+ * The fields on this interface are nullable because typically the spec will _not_ have that information available
+ * before the corresponding [ResourceHandler] has resolved the resource.
+ */
+interface ArtifactProvider {
+  val artifactName: String?
+  val artifactType: ArtifactType?
+
+  fun completeArtifactOrNull() =
+    if (artifactName != null && artifactType != null) {
+      CompleteArtifact(artifactName!!, artifactType!!)
+    } else {
+      null
+    }
+}
+
+/**
+ * Implemented by [ResourceSpec] or concrete resource types that (may) contain versioned artifacts, typically compute
+ * resources.
+ */
+interface VersionedArtifactProvider : ArtifactProvider {
+  val artifactVersion: String?
+
+  fun completeVersionedArtifactOrNull() =
+    if (artifactName != null && artifactType != null && artifactVersion != null) {
+      CompleteVersionedArtifact(artifactName!!, artifactType!!, artifactVersion!!)
+    } else {
+      null
+    }
+}
+
+/**
+ * Implemented by [ResourceSpec] or concrete resource types that (may) contain artifact references, typically compute
+ * resources.
+ */
+interface ArtifactReferenceProvider {
+  val artifactReference: String?
```

Comment:
As someone who's still learning the codebase, I'm struggling to understand what kind of information is encoded in the artifactReference string. Can you provide an example of an artifact reference?

Reply:
It's just an arbitrary string you can assign to an artifact so it can be referenced elsewhere in the delivery config. Here's an example:
https://github.com/spinnaker/keel/blob/0c33aab98a45e679e3833c5baf5f2d58b37682d5/keel-web/src/test/resources/examples/delivery-config-example.yml#L5-L8
Which gets referenced here:
https://github.com/spinnaker/keel/blob/0c33aab98a45e679e3833c5baf5f2d58b37682d5/keel-web/src/test/resources/examples/delivery-config-example.yml#L31-L34

Reply:
Oh, it's an identifier!

Reply:
yep, just a friendly string identifier :)